python/ray/data/aggregate.py

```diff
@@ -71,17 +74,127 @@ def __init__(self):
+        r: This row.
+
+    Returns:
+        Accumulated result including the provided row
+    """
+    r = on_fn(r)
+    if _is_null(r):
+        if ignore_nulls:
+            return a
+        else:
+            return None
+    else:
+        if a == _INIT_SENTINEL:
+            return first(r)
+        elif a is None:
+            return None
+        else:
+            return accum(a, r)
+
+
+def _null_merge(
+    ignore_nulls: bool,
+    merge: Callable[[AggType, AggType], AggType],
+    a1: AggType,
+    a2: AggType,
+) -> AggType:
+    """
+    Merge two accumulations while handling nulls.
+
+    This merges two accumulations subject to the following null rules:
+    1. If a1 is empty and a2 is empty, return empty sentinel.
+    2. If a1 (a2) is empty and a2 (a1) is null, return null.
+    3. If a1 (a2) is empty and a2 (a1) is non-null, return a2 (a1).
+    4. If a1 (a2) is null, return a2 (a1) if ignoring nulls, null otherwise.
+    5. If a1 and a2 are both non-null, return merge(a1, a2).
+
+    Args:
+        ignore_nulls: Whether nulls should be ignored or cause a null result.
+        merge: The core merge function.
+        a1: One of the intermediate accumulations to merge.
+        a2: One of the intermediate accumulations to merge.
+
+    Returns:
+        Accumulation of the two provided accumulations.
+    """
+    if a1 == _INIT_SENTINEL:
+        # If one operand is the init value, propagate the other.
+        # No matter whether a2 is a real value, the init value, or None,
+        # propagating each of these is correct if a1 is the init value.
+        return a2
+    if a1 is None:
+        # If we're ignoring nulls, propagate a2; otherwise, propagate null.
+        return a2 if ignore_nulls else None
+    if a2 == _INIT_SENTINEL:
+        # If one operand is the init value, propagate the other.
+        return a1
+    if a2 is None:
+        # If we're ignoring nulls, propagate a1; otherwise, propagate null.
+        return a1 if ignore_nulls else None
+    return merge(a1, a2)
+
+
+def _null_finalize(finalize: Callable[[AggType], AggType], a: AggType) -> AggType:
+    """
+    Finalize an accumulation while handling nulls.
+
+    If the accumulation is empty or null, this returns null.
+
+    Args:
+        finalize: The core finalizing function.
+        a: Accumulation result to finalize.
+
+    Returns:
+        Finalized accumulation result.
+    """
+    return finalize(a) if a != _INIT_SENTINEL and a is not None else None
+
+
 @PublicAPI(stability="beta")
 class Sum(_AggregateOnKeyBase):
     """Defines sum aggregation."""
 
-    def __init__(self, on: Optional[KeyFn] = None):
+    def __init__(self, on: KeyFn = None, ignore_nulls: bool = True):
         self._set_key_fn(on)
         on_fn = _to_on_fn(on)
+
         super().__init__(
-            init=lambda k: 0,
-            accumulate=lambda a, r: a + on_fn(r),
-            merge=lambda a1, a2: a1 + a2,
+            init=lambda k: _INIT_SENTINEL,
```

Comment:
Agreed that we need a better empty sentinel here.

The core issue is that our policy for merging two accumulations differs depending on whether an accumulation is `None` or empty (sentinel):
- `None`: propagate the other accumulation if we're ignoring nulls, otherwise return None
- Empty: always propagate the other accumulation

So you're suggesting using `[/* init data */] + [bool(is_empty)]` as the initial aggregation... That could be nicer, since we could go back to always feeding the accumulator the proper per-aggregation init data and we can get rid of the `first` callable in `null_accumulate`.

Let me make that change!

Reply:
Hmm so we can't have mixed types in the accumulation rows since it breaks the subsequent Arrow block...

Reply:
what about change bool -> int/float as well?